Update new user groups UI to be more like the streams UI.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- FIrst four commits are preparatory refactors and fixes.
- Next two commits are to make joining and leaving groups easy.
- Next commit is to show a message when groups list is empty.
- Next commit is to reset the right panel, if required, when toggling between lists in left panel.
- Next commit is to show loading spinner in place of checkmark and plus icon.
- Next commit is to disable "Add" button for members list in case of empty inputs.
- Last commit is to add support for changing `can_mention_group` setting.

Fixes: <!-- Issue link, or clear description.--> #25538

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**GIFS**
![not-allowed-to join-groups](https://github.com/zulip/zulip/assets/35494118/f481b47d-df4d-4f98-b7ca-a4694ea3dc1f)
![group-settings](https://github.com/zulip/zulip/assets/35494118/bb0d18c4-1438-41aa-862c-d64677c80f45)
![group-creation](https://github.com/zulip/zulip/assets/35494118/54a52019-4e09-4de1-a7c8-ba0708a320f8)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on web/src/user_group_edit.js:
```diff
@@ -303,6 +304,33 @@ export function setup_group_list_tab_hash(tab_key_value) {
+    /* Destroy the spinner. */
+    const $spinner = $(group_row).find(".join_leave_status").expectOne();
+    loading.destroy_indicator($spinner);
+}
```
This leads to some duplication. We can probably move these functions to `settings_components` and keep only single function.

alya commented:
1. Let's change: You are not part of any user group. -> You are not a member of any user groups.
2. Let's drop this tooltip; I don't think it's helpful, and the button is clear enough on its own.
![Screenshot 2023-11-27 at 9 46 39 AM](https://github.com/zulip/zulip/assets/2090066/153afe12-80ef-409c-bee7-bebe26cdabcb)
3. The "no results" string need to be updated... Probably something like:
   - No members at all: "This group has no members."
   - No members with a filter: "No group members match your current filter." (worded to match Users panel)

![Screenshot 2023-11-27 at 9 47 53 AM](https://github.com/zulip/zulip/assets/2090066/88e8aef2-7f0b-41d9-b8aa-6dedbe12de91)

4. Also, the strings in the streams UI should be updated:
   - No members at all: "This stream has no subscribers."
   - No members with a filter: "No stream subscribers match your current filter."
5. The button to remove someone from a group should be labeled "Remove", not "Unsubscribe".

![Screenshot 2023-11-27 at 9 49 05 AM](https://github.com/zulip/zulip/assets/2090066/9387ee08-fd32-44cf-b903-c10d2758eafc)

alya commented:
6. I don't mind this being a follow-up, but the "Actions" table heading gets squashed to the right when there are no stream members / group subscribers.

 
![Screenshot 2023-11-27 at 9 54 18 AM](https://github.com/zulip/zulip/assets/2090066/89c0132c-f925-4027-99d9-22a759634828)

alya commented:
That's all I've got right now!

timabbott commented:
I did a read and this basically looks good to me. I think releasing the `can_mention_user_group` feature is OK; we've done the important part of implementing that how we want for system groups.

timabbott commented:
I think these strings might be better written as, e.g., `Join group {{name}}` without spelling out "user group", since in the context it's pretty clear we're talking about you as a user? 

```
Cannot join {{name}} user group
Cannot leave {{name}} user group
Join {{name}} user group
Leave {{name}} user group
```

sahil839 commented:
Made all the changes except to show the different messages when users list is empty because using the existing code to do this results in different styling of the message, which I am not sure is the best.
![Screenshot from 2023-11-28 09-36-29](https://github.com/zulip/zulip/assets/35494118/564a6b21-82af-473b-b5f8-0de302547f1b)


For now have just updated the text as mentioned above but same text is shown for the case when filter results is empty -
![Screenshot from 2023-11-28 09-39-18](https://github.com/zulip/zulip/assets/35494118/dc44da26-ed8d-4f16-90a4-ef76bcca0a72)


![Screenshot from 2023-11-28 09-39-52](https://github.com/zulip/zulip/assets/35494118/18189575-5a6e-44aa-aeeb-60679a8471b2)

alya commented:
We should definitely use different strings when the filter has content in it; otherwise, the string is just incorrect in that situation.

sahil839 commented:
Updated the PR with following changes -
- Updated the text to "You are not a member of any user groups.".
- Removed the tooltip from join/leave button.
- Renamed the "Unsubscribe" button to "Remove" in user group members list.
- Updated the text shown for empty subscribers and members list.
- Updated the tooltip text for the checkmark in left section.
- Also added commit to remove some unnecessary CSS and also added bottom border for the table such that the new styling for empty message in table looks somewhat fine. Will share the screenshots below.

alya commented:
Looks like https://github.com/zulip/zulip/pull/27912#issuecomment-1828341257 has been fixed for the new group UI, but not for existing groups.

alya commented:
That's all I've got right now. I think neither of these is a blocker for merging this PR, if that simplifies matters.

timabbott commented:
Those seem like good follow-ups, since the implementation changes here all look like good improvements, and I just got through a review without changes I wanted to suggest at the code level. Merged, thanks @sahil839!
